### PR TITLE
add install generator for rails

### PR DESCRIPTION
**Pull request by vitalikdanchenko**

### CODE REVIEW

File: lib/talkable/generators/install_generator.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+module Talkable
+  class InstallGenerator < Rails::Generators::Base
+    source_root File.expand_path("../templates", __FILE__)
+    class_option :haml, type: :boolean, default: false
+
+    def ask_config_values
+      @site_slug  = ask("Your Talkable site slug:")
+      @api_key    = ask("Your Talkable API Key:")
+      if ['', 'y', 'Y'].include?(ask 'Do you have a custom domain? [Y/n]')
+        @server   = ask("Your custom [#{Talkable::Configuration::DEFAULT_SERVER}]:")
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb]: Your custom what?

File: lib/talkable/generators/install_generator.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      end
+    end
+
+    def add_initializer
+      template "config/initializers/talkable.rb", "config/initializers/talkable.rb"
+    end
+
+    def inject_talkable_offer
+      inject_into_file "app/controllers/application_controller.rb", :after => "protect_from_forgery with: :exception" do
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb]: 1. use modern syntax please
2. what will happen if  `protect_from_forgery with: :exception` is absent?

File: lib/talkable/generators/install_generator.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+RUBY
+      end
+
+      inject_into_file "app/controllers/application_controller.rb", :before => /^end/ do
+<<-RUBY
+  protected
+
+  def load_talkable_offer
+    origin = Talkable.register_affiliate_member(campaign_tags: 'popup')
+    @offer ||= origin&.offer
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb]: We support old rubies. Can we add if condition here?

File: lib/talkable/generators/install_generator.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    def inject_talkable_offer
+      inject_into_file "app/controllers/application_controller.rb", :after => "protect_from_forgery with: :exception" do
+<<-RUBY
+
+  before_action :load_talkable_offer
+RUBY
+      end
+
+      inject_into_file "app/controllers/application_controller.rb", :before => /^end/ do
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb]: Use modern syntax

File: lib/talkable/generators/install_generator.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+<<-RUBY
+  protected
+
+  def load_talkable_offer
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb:31]: I think it would be better use 

```
protected def load_talkable_offer
  ...
end
```

Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb:31]: But old rubies.....

File: lib/talkable/generators/install_generator.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    origin = Talkable.register_affiliate_member(campaign_tags: 'popup')
+    @offer ||= origin&.offer
+  end
+RUBY
+      end
+
+      empty_directory "app/views/shared/"
+
+      if options[:haml]
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb]: let's support slim too

File: lib/talkable/generators/templates/config/initializers/talkable.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+Talkable.configure do |config|
+  # site slug takes form ENV["TALKABLE_SITE_SLUG"]
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/templates/config/initializers/talkable.rb:2]: `Site slug is taken from ENV["TALKABLE_SITE_SLUG"]`

File: lib/talkable/generators/install_generator.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      if ['', 'y', 'Y'].include?(ask 'Do you have a custom domain? [Y/n]')
+        @server   = ask("Your custom domain [#{Talkable::Configuration::DEFAULT_SERVER}]:")
+      end
+    end
+
+    def add_initializer
+      template "config/initializers/talkable.rb", "config/initializers/talkable.rb"
+    end
+
+    def inject_talkable_offer
```
Comment (Tonkonozhenko) [inline lib/talkable/generators/install_generator.rb:19]: What about such syntax?

``` ruby
    def inject_talkable_offer
      inject_into_file "app/controllers/application_controller.rb", <<-RUBY, after: "protect_from_forgery with: :exception"

  before_action :load_talkable_offer
      RUBY

      ...
```

Comment (zhuravel) [reply lib/talkable/generators/install_generator.rb:19]: Personally I don't like it and have a hard time reading it every time.

Comment (Tonkonozhenko) [reply lib/talkable/generators/install_generator.rb:19]: I don't like when there are more than one heredocs

``` ruby
x(<<DOC1, <<DOC2, <<DOC3)
  wow
DOC1
  so much
DOC2
  heredocs
DOC3
```

**Comment by Tonkonozhenko** [comment]
LG

File: lib/talkable/generators/install_generator.rb (hunk 8 above)
Comment (vitalikdanchenko) [reply lib/talkable/generators/install_generator.rb:19]: I don't like either, sometime it looks prettier but most of all it only confuses.

File: lib/talkable/generators/install_generator.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  def load_talkable_offer
+    origin = Talkable.register_affiliate_member(campaign_tags: 'popup')
+    @offer ||= origin.offer if origin
+  end
+RUBY
+      end
+
+      empty_directory "app/views/shared/"
+
+      if options[:haml] || options[:slim]
```
Comment (bogdan) [inline lib/talkable/generators/install_generator.rb:40]: I think you can guess which template engine to use by looking at the extension of file like `app/views/layouts/application.html.xxx`

**Comment by bogdan** [comment]
It seems we don't provide an access to controller that controls `/invite` and the view. This is bad in general app as at some moment you will want to have it to control layout, access permission etc.

File: lib/talkable/generators/install_generator.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  class InstallGenerator < Rails::Generators::Base
+    source_root File.expand_path("../templates", __FILE__)
+    class_option :haml, type: :boolean, default: false
+    class_option :slim, type: :boolean, default: false
+
+    def ask_config_values
+      @site_slug  = ask("Your Talkable site slug:")
+      @api_key    = ask("Your Talkable API Key:")
+      if ['', 'y', 'Y'].include?(ask 'Do you have a custom domain? [Y/n]')
+        @server   = ask("Your custom domain [#{Talkable::Configuration::DEFAULT_SERVER}]:")
```
Comment (bogdan) [inline lib/talkable/generators/install_generator.rb:11]: There can be a special method for `y/n` question.

File: lib/talkable/generators/install_generator.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      end
+    end
+
+    def add_initializer
+      template "config/initializers/talkable.rb", "config/initializers/talkable.rb"
+    end
+
+    def inject_talkable_offer
+      inject_into_file "app/controllers/application_controller.rb", after: "protect_from_forgery with: :exception" do
```
Comment (bogdan) [inline lib/talkable/generators/install_generator.rb:20]: What is a purpose of it?

File: lib/talkable/generators/install_generator.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+<<-RUBY
+
+  before_action :load_talkable_offer
+RUBY
+      end
+
+      inject_into_file "app/controllers/application_controller.rb", before: /^end/ do
+<<-RUBY
+  protected
```
Comment (bogdan) [inline lib/talkable/generators/install_generator.rb:29]: I am pretty much sure `protected` would be already in a file and we can skip it.

**Comment by vitalikdanchenko** [comment]
`invite` campaign is in a process, we have only `trigger widget` here

File: lib/talkable/generators/install_generator.rb (hunk 12 above)
Comment (vitalikdanchenko) [reply lib/talkable/generators/install_generator.rb:29]: `application_controller.rb` in new generated application doesn't have it

File: lib/talkable/generators/install_generator.rb (hunk 11 above)
Comment (vitalikdanchenko) [reply lib/talkable/generators/install_generator.rb:20]: trigger widget for all pages

**Pull request merged**